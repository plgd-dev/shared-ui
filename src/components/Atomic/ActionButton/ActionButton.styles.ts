import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import { get, ThemeType } from '../_theme'

export const actionButton = css`
    width: 20px;
    height: 20px;
`

export const floatingMenu = (theme: ThemeType) => css`
    z-index: 10;
    background: ${get(theme, `ActionButton.floatingMenu.background`)};
    border: 1px solid ${get(theme, `ActionButton.floatingMenu.borderColor`)};
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    padding: 8px 16px;
`

export const item = (theme: ThemeType) => css`
    display: flex;
    align-items: center;
    text-decoration: none;
    color: ${colors.neutral500};
    transition: all 0.3s;
    padding: 8px 0;
    cursor: pointer;
    border-bottom: 1px solid ${get(theme, `ActionButton.item.borderColor`)};

    &:hover {
        color: ${get(theme, `ActionButton.item.hover.color`)};
    }

    &:last-child {
        border-bottom: 0;
    }
`

export const itemIcon = css`
    margin-right: 10px;
`

export const itemLabel = css`
    font-size: 14px;
    line-height: 14px;
`

export const icon = (theme: ThemeType) => css`
    width: 20px;
    height: 20px;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${colors.neutral500};
    transition: all 0.3s;
    cursor: pointer;

    &:hover {
        color: ${get(theme, `ActionButton.icon.hover.color`)};
    }
`

export const iconActive = (theme: ThemeType) => css`
    color: ${get(theme, `ActionButton.icon.acive.color`)};
`
