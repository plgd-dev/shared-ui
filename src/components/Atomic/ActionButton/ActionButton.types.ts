import { ReactNode, SyntheticEvent } from 'react'
import { dropdownTypes } from './constants'
import { Placement } from '@floating-ui/react-dom-interactions'

export type ActionButtonTypeType = (typeof dropdownTypes)[keyof typeof dropdownTypes]

export type ActionButtonItemType = {
    component?: ReactNode
    hidden?: boolean
    icon?: ReactNode
    id?: string
    label: string
    loading?: boolean
    onClick?: () => void
}

export type Props = {
    className?: string
    dataTestId?: string
    dataTestIdDropdown?: string
    defaultOpen?: boolean
    id?: string
    items: ActionButtonItemType[]
    loading?: boolean
    menuProps?: {
        placement?: Placement
    }
    onToggle?: (isOpen: boolean, event: SyntheticEvent<any>) => void
    portalTarget?: ReactNode | Element | null
    type?: ActionButtonTypeType
    disabled?: boolean
}

export const defaultProps = {
    type: dropdownTypes.EMPTY,
}
