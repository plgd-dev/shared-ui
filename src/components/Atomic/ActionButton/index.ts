export { default } from './ActionButton'

export * from './constants'
