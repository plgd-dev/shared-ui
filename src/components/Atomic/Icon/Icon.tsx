import { FC, forwardRef, SVGProps, useEffect, useRef, useState } from 'react'
import { Props, defaultProps } from './Icon.types'
import * as styles from './Icon.styles'

export const Icon = forwardRef<any, Props>((props, ref) => {
    const { className, icon, id, size, onError, onCompleted, ...rest } = { ...defaultProps, ...props }

    const [loading, setLoading] = useState(false)
    const [iconLoad, setIconLoad] = useState('')
    const [error, setError] = useState<Error | undefined>(undefined)
    const ImportedIconRef = useRef<FC<SVGProps<SVGSVGElement>> | undefined>(undefined)

    const IconComponent = ImportedIconRef.current ? ImportedIconRef.current : () => null

    const parseComponentName = (string: string) =>
        string
            .split('-')
            .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
            .join('')

    useEffect(() => {
        setLoading(true)
        const componentName = parseComponentName(icon)
        const importIcon = async (): Promise<void> => {
            try {
                await import(`./components/${componentName}.tsx`).then((IconComponent) => {
                    ImportedIconRef.current = IconComponent.default
                    return IconComponent.default
                })
                onCompleted?.(icon, ImportedIconRef.current)
            } catch (err: any) {
                onError?.(err)
                setError(err)
            } finally {
                setLoading(false)
            }
        }
        importIcon().then(() => {
            setIconLoad(icon)
        })
    }, [icon, onCompleted, onError])

    if (error) {
        console.error(error)
        return null
    }

    return (
        <div
            className={className}
            css={styles.icon}
            data-icon={iconLoad}
            id={id}
            ref={ref}
            style={{
                width: size,
                height: size,
            }}
        >
            {loading ? null : <IconComponent {...rest} height={size} width={size} />}
        </div>
    )
})

Icon.displayName = 'Icon'

export default Icon
