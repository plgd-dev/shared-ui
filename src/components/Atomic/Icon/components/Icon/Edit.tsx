// Auto-generated file created by svgr-cli source svg-template.js
// Run yarn icons:create to update
// Do not edit
import * as React from 'react'
import { SVGProps } from 'react'
const SvgEdit = (props: SVGProps<SVGSVGElement>) => (
    <svg width={props.width || 16} height={props.height || 16} viewBox='0 0 20 20' fill='none' xmlns='http://www.w3.org/2000/svg' {...props}>
        <path
            d='M17.178 2.207a1.84 1.84 0 0 0-2.605 0l-.99.987 3.22 3.22.99-.99a1.84 1.84 0 0 0 0-2.605l-.615-.612Zm-9.84 7.236c-.202.2-.356.447-.445.72l-.974 2.92a.793.793 0 0 0 .191.81c.21.214.523.286.81.19l2.92-.973c.27-.092.517-.243.72-.444l5.503-5.506-3.223-3.223-5.503 5.506ZM4.823 3.598a3.158 3.158 0 0 0-3.157 3.158v8.42a3.158 3.158 0 0 0 3.157 3.157h8.42a3.158 3.158 0 0 0 3.158-3.157v-3.158a1.051 1.051 0 1 0-2.105 0v3.158c0 .582-.47 1.052-1.053 1.052h-8.42c-.582 0-1.052-.47-1.052-1.052v-8.42c0-.582.47-1.053 1.052-1.053h3.158a1.051 1.051 0 1 0 0-2.105H4.824Z'
            fill='#81868C'
        />
    </svg>
)
export default SvgEdit
