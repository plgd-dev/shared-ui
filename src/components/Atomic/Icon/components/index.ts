export { default as IconActions } from './IconActions'
export { default as IconArrowDetail } from './IconArrowDetail'
export { default as IconArrowDownNoPadding } from './IconArrowDownNoPadding'
export { default as IconArrowDown } from './IconArrowDown'
export { default as IconArrowLeft } from './IconArrowLeft'
export { default as IconArrowRight } from './IconArrowRight'
export { default as IconArrowTriangleFullUp } from './IconArrowTriangleFullUp'
export { default as IconBell } from './IconBell'
export { default as IconCertificate } from './IconCertificate'
export { default as IconChat } from './IconChat'
export { default as IconCloseCircle } from './IconCloseCircle'
export { default as IconClose } from './IconClose'
export { default as IconCloudError } from './IconCloudError'
export { default as IconCloudSuccess } from './IconCloudSuccess'
export { default as IconCloudWarning } from './IconCloudWarning'
export { default as IconCollapse } from './IconCollapse'
export { default as IconCopy } from './IconCopy'
export { default as IconDashboard } from './IconDashboard'
export { default as IconDeviceUpdate } from './IconDeviceUpdate'
export { default as IconDevices } from './IconDevices'
export { default as IconDocs } from './IconDocs'
export { default as IconDownload } from './IconDownload'
export { default as IconEdit } from './IconEdit'
export { default as IconError } from './IconError'
export { default as IconFilePem } from './IconFilePem'
export { default as IconFileUpload } from './IconFileUpload'
export { default as IconGlobe } from './IconGlobe'
export { default as IconHidePassword } from './IconHidePassword'
export { default as IconInfo } from './IconInfo'
export { default as IconIntegrations } from './IconIntegrations'
export { default as IconLink } from './IconLink'
export { default as IconLoader } from './IconLoader'
export { default as IconLock } from './IconLock'
export { default as IconLog } from './IconLog'
export { default as IconLogout } from './IconLogout'
export { default as IconNet } from './IconNet'
export { default as IconNetwork } from './IconNetwork'
export { default as IconNotification } from './IconNotification'
export { default as IconPendingCommands } from './IconPendingCommands'
export { default as IconPlus } from './IconPlus'
export { default as IconRefresh } from './IconRefresh'
export { default as IconRemoteClients } from './IconRemoteClients'
export { default as IconSearch } from './IconSearch'
export { default as IconSettings } from './IconSettings'
export { default as IconShield } from './IconShield'
export { default as IconShowPassword } from './IconShowPassword'
export { default as IconSortDown } from './IconSortDown'
export { default as IconSortUp } from './IconSortUp'
export { default as IconSuccess } from './IconSuccess'
export { default as IconTableArrowDown } from './IconTableArrowDown'
export { default as IconTableArrowUp } from './IconTableArrowUp'
export { default as IconToastClose } from './IconToastClose'
export { default as IconTrash } from './IconTrash'
export { default as IconWarning } from './IconWarning'
