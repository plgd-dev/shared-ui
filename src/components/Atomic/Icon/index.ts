export { default } from './Icon'
export * from './Icon'

export * from './components'

export * from './utils'
