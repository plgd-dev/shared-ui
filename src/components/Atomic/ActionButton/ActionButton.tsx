import React, { cloneElement, FC, ReactElement, useEffect, useRef, useState } from 'react'
import { ActionButtonItemType, defaultProps, Props } from './ActionButton.types'
import { convertSize, Icon, IconActions } from '../Icon'
import * as styles from './ActionButton.styles'
import isFunction from 'lodash/isFunction'
import { offset, shift, useFloating } from '@floating-ui/react'
import { createPortal } from 'react-dom'

const ActionButton: FC<Props> = (props) => {
    const { menuProps, items, onToggle, className, id, defaultOpen, loading, dataTestIdDropdown, portalTarget } = props
    const [open, setOpen] = useState(defaultOpen)
    const ref = useRef(null)
    const { x, y, reference, floating, strategy } = useFloating({
        placement: menuProps?.placement || 'bottom-end',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })

    useEffect(() => {
        function handleClickOutside(event: any) {
            // @ts-ignore
            if (ref?.current && !ref?.current?.contains(event.target) && event.target.getAttribute('data-close-dropdown') === false) {
                setOpen(false)
                isFunction(onToggle) && onToggle(false)
            }
        }
        if (open) {
            document.addEventListener('mousedown', handleClickOutside)
            return () => {
                document.removeEventListener('mousedown', handleClickOutside)
            }
        }
    }, [ref, open, onToggle])

    const handleOpen = () => {
        if (!loading) {
            const newState = !open
            setOpen(newState)
            isFunction(onToggle) && onToggle(newState)
        }
    }

    const getIcon = (item: ActionButtonItemType) => {
        if (item.icon) {
            return typeof item.icon === 'string' ? (
                <Icon css={styles.itemIcon} data-close-dropdown={false} icon={item.icon} size={20} />
            ) : (
                cloneElement(item.icon as ReactElement, { ...convertSize(20), 'css': styles.itemIcon, 'data-close-dropdown': false })
            )
        }

        return null
    }

    const floatingPanel = (
        <div
            css={styles.floatingMenu}
            ref={floating}
            style={{
                position: strategy,
                top: y ?? 0,
                left: x ?? 0,
                width: 'max-content',
            }}
        >
            {items
                .filter((item) => !item.hidden)
                .map((item) => {
                    return (
                        item.component || (
                            <div
                                css={styles.item}
                                data-close-dropdown={false}
                                key={item.id || item.label}
                                onClick={(e) => {
                                    open && setOpen(false)
                                    isFunction(onToggle) && onToggle(false)
                                    isFunction(item.onClick) && item.onClick(e)
                                }}
                            >
                                {getIcon(item)}
                                <span css={styles.itemLabel} data-close-dropdown={false}>
                                    {item.label}
                                </span>
                            </div>
                        )
                    )
                })}
        </div>
    )

    return (
        <div className={className} css={styles.actionButton} id={id} ref={ref}>
            <div css={[styles.icon, open && styles.iconActive]} data-test-id={dataTestIdDropdown} onClick={handleOpen} ref={reference}>
                <IconActions {...convertSize(20)} />
            </div>

            {open && portalTarget && createPortal(floatingPanel, portalTarget as Element)}
            {open && !portalTarget && floatingPanel}
        </div>
    )
}

ActionButton.displayName = 'ActionButton'
ActionButton.defaultProps = defaultProps

export default ActionButton
