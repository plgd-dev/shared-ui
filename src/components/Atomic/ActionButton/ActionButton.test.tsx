import { waitFor } from '@testing-library/react'

import { render } from '../../../../test/jest-wrapper'
import ActionButton from './ActionButton'
import { IconEdit, IconPlus, IconTrash } from '../Icon'

describe('<ActionButton>', () => {
    it('render correctly - snapshot', async () => {
        const { asFragment } = render(
            <ActionButton
                defaultOpen={true}
                items={[
                    {
                        onClick: () => console.log('Create'),
                        label: 'Create',
                        icon: <IconPlus />,
                    },
                    {
                        onClick: () => console.log('Update'),
                        label: 'Update',
                        icon: <IconEdit />,
                    },
                    {
                        onClick: () => console.log('Delete'),
                        label: 'Delete',
                        icon: <IconTrash />,
                    },
                ]}
            />
        )

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot()
        })
    })
})
