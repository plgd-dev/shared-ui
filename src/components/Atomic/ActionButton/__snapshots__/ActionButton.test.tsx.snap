// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionButton> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 20px;
  height: 20px;
}

.emotion-1 {
  width: 20px;
  height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #81868C;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  cursor: pointer;
  color: #007BBF;
}

.emotion-1:hover {
  color: #007BBF;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-3 {
  z-index: 10;
  background: #fff;
  border: 1px solid #e6e9ed;
  box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
  border-radius: 8px;
  padding: 8px 16px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #81868C;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  padding: 8px 0;
  cursor: pointer;
  border-bottom: 1px solid #E6E9ED;
}

.emotion-4:hover {
  color: #2261AE;
}

.emotion-4:last-child {
  border-bottom: 0;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  margin-right: 10px;
}

.emotion-6 {
  font-size: 14px;
  line-height: 14px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
        data-icon=""
        style="width: 20px; height: 20px;"
      />
    </div>
    <div
      class="emotion-3"
      style="position: fixed; top: 0px; left: 0px;"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
          data-icon=""
          style="width: 20px; height: 20px;"
        />
        <span
          class="emotion-6"
        >
          Create
        </span>
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
          data-icon=""
          style="width: 20px; height: 20px;"
        />
        <span
          class="emotion-6"
        >
          Update
        </span>
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
          data-icon=""
          style="width: 20px; height: 20px;"
        />
        <span
          class="emotion-6"
        >
          Delete
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
